### PR TITLE
Add end-to-end tests for deprecation

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/nuget/nugetgallery/issues/6852

I've added a tests that deprecates a package.

Had to make some minor changes to the infrastructure to fetch and poll deprecation metadata as well.

### CODE REVIEW

**Comment by joelverhagen** [comment]
What is the test suite duration before and after adding these tests?

**Comment by scottbommarito** [comment]
@joelverhagen - I thought these tests ran in parallel so I added more tests than necessary, which bloated the runtime substantially. I have reduced the set of tests and will determine the new impact on runtime. I expect it to be in the 5-10 min range.

**Comment by skofman1** [comment]
"UndeprecatePackage" test is not critical. If it increases the run duration consider dropping it.

**Comment by scottbommarito** [comment]
@skofman1 - Removed the test after our offline discussion.

**Comment by scottbommarito** [comment]
Without the undeprecate test, this change has no impact on E2E runtime.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 1]
```diff
@@ -157,10 +160,32 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
             await SendToPackageAsync(HttpMethod.Delete, id, version, logger);
         }
 
+        public async Task DeprecateAsync(string id, IReadOnlyCollection<string> versions, PackageDeprecationContext context, ITestOutputHelper logger)
+        {
+            var galleryEndpoint = await GetGalleryUrlAsync(logger);
+            var url = $"{galleryEndpoint}/api/v2/package/{id}/deprecations";
+
+            var bodyJObject = new JObject(
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: nit: `JsonConvert` on an anonymous type would be more readable

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs [hunk 2]
```diff
@@ -60,6 +59,61 @@ public RegistrationClient(SimpleHttpClient httpClient, V3IndexClient v3IndexClie
...
+
+            await PollAsync(
+                id,
+                version,
+                semVer2: true,
+                isComplete: catalogEntry => 
+                    catalogEntry.Id == id 
+                    && catalogEntry.Version == version
+                    && (catalogEntry.Deprecation == null) == (deprecation == null)
+                    && HasReason(catalogEntry, "Legacy", deprecation?.IsLegacy ?? false)
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: Failures will be a lot more debuggable if this message returned the deprecation object and the test runs assertions

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs [hunk 3]
```diff
@@ -60,6 +59,61 @@ public RegistrationClient(SimpleHttpClient httpClient, V3IndexClient v3IndexClie
...
+                    && HasReason(catalogEntry, "CriticalBugs", deprecation?.HasCriticalBugs ?? false)
+                    && HasReason(catalogEntry, "Other", deprecation?.IsOther ?? false)
+                    && deprecation?.Message == catalogEntry.Deprecation?.Message 
+                    && deprecation?.AlternatePackageId == catalogEntry.Deprecation?.AlternatePackage?.Id
+                    && catalogEntry.Deprecation?.AlternatePackage?.Range == GetExpectedAlternatePackageRange(deprecation),
+                startingMessage: $"Waiting for package {id} {version} to be {successState} on registration base URLs:",
+                successMessageFormat: $"Package {id} {version} became {successState} on {{0}} after waiting {{1}}.",
+                failureMessageFormat: $"Package {id} {version} was still {failureState} on {{0}} after waiting {{1}}.",
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: If one of the deprecation properties (e.g. the message) does not match but the package is still deprecated `failureMessageFormat` will say something confusing.

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs [hunk 4]
```diff
@@ -60,6 +59,61 @@ public RegistrationClient(SimpleHttpClient httpClient, V3IndexClient v3IndexClie
...
 
+        /// <summary>
+        /// Polls all registration base URLs until the provided ID and version has the specified listed state.
+        /// </summary>
+        /// <param name="id">The package ID.</param>
+        /// <param name="version">The package version.</param>
+        /// <param name="deprecation">The deprecation state to poll for.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns>Returns a task that completes when the package has the specified deprecation state or the timeout has occurred.</returns>
+        public async Task WaitForDeprecationStateAsync(
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: IMO this polling is too complex. Perhaps just check the deprecation is non-null and let the caller do the nitty gritty

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs [hunk 5]
```diff
@@ -60,6 +59,61 @@ public RegistrationClient(SimpleHttpClient httpClient, V3IndexClient v3IndexClie
                 logger: logger);
         }
 
+        /// <summary>
+        /// Polls all registration base URLs until the provided ID and version has the specified listed state.
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: `deprecation state`

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs [hunk 6]
```diff
@@ -238,6 +292,20 @@ public class CatalogEntry
             public string LicenseExpression { get; set; }
             public string LicenseUrl { get; set; }
             public string IconUrl { get; set; }
+            public CatalogDeprecation Deprecation { get; set; }
+        }
+
+        public class CatalogDeprecation
+        {
+            public string Message { get; set; }
+            public IEnumerable<string> Reasons { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: nit: List<string>/IReadOnlyList<string> for easier indexing and clear enumeration

Comment (scottbommarito) [reply src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: done

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs (hunk 4 above)
Comment (scottbommarito) [reply src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: sounds good.

File: src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs]: makes sense--changed it

**Comment by scottbommarito** [comment]
Waiting to merge this until https://github.com/NuGet/NuGetGallery/commit/9c3410ed69d1650484322a9a4bafcf544d919ab6 goes into PROD.

**Comment by scottbommarito** [comment]
screwed up my rebase to pull in latest master--will fix

**Pull request merged**